### PR TITLE
Update instructions for running XDP example

### LINKED ISSUE #14: Update XDP example with missing dependencies and instruction to run
Update XDP example with missing dependencies and instruction to run

### PR DESCRIPTION
**Pull request by nitishm**
Closes #14 

- Logging Packet document instructions updated with `lo` (localhost)
interface as the example since eth0 may not be available on all machines
- Update template files for the example application with missing
dependencies and import statement. Switch from eth0 interface to lo
interface

Signed-off-by: Nitish Malhotra <nitish.malhotra@gmail.com>

### CODE REVIEW

**Comment by nitishm** [comment]
@dave-tucker @alessandrod PTAL

**Review by alessandrod — COMMENTED** [review]
This looks good to me, but I'll leave it to @dave-tucker to approve (I'm not sure if not using structopt was intentional)

File: examples/myapp-03/myapp-ebpf/src/main.rs [hunk 1]
```diff
@@ -68,18 +68,18 @@ fn try_xdp_firewall(ctx: XdpContext) -> Result<u32, ()> {
     if h_proto != ETH_P_IP {
         return Ok(xdp_action::XDP_PASS);
     }
-    let source = u32::from_be(unsafe { *ptr_at(&ctx, ETH_HDR_LEN + offset_of!(iphdr, saddr))? });
+    let daddr = u32::from_be(unsafe { *ptr_at(&ctx, ETH_HDR_LEN + offset_of!(iphdr, daddr))? });
```
Comment (dave-tucker) [inline examples/myapp-03/myapp-ebpf/src/main.rs:71]: changing from source to destinateion address will require updating chapter 2 too.

File: examples/myapp-03/myapp/src/main.rs [hunk 2]
```diff
@@ -1,38 +1,41 @@
...
+    let opt = Opt::from_args();
+    let mut bpf = Bpf::load_file(&opt.path)?;
+    let program: &mut Xdp = bpf.program_mut("myapp")?.try_into()?;
+    program.load()?;
+    program.attach(&opt.iface, XdpFlags::default())?;
 
     // ANCHOR: block_address
     let mut blocklist: HashMap<_, u32, u32> = HashMap::try_from(bpf.map_mut("BLOCKLIST")?)?;
-    let block_addr : u32 = Ipv4Addr::new(1, 1, 1, 1).try_into()?;
+    // Let's block localhost IP addresses
```
Comment (dave-tucker) [inline examples/myapp-03/myapp/src/main.rs:37]: 192.168.0.10 isn't localhost so this comment doesn't make sense

File: src/start/dropping-packets.md [hunk 3]
```diff
@@ -61,18 +61,26 @@ We'll use the `IPv4Addr` type to represent our IP address as it's human-readable
 ```console
 cargo build
 cargo xtask build-ebpf
-sudo ./target/debug/myapp ./target/bpfel-unknown-none/debug/myapp wlp2s0
+sudo ./target/debug/myapp --path ./target/bpfel-unknown-none/debug/myapp --intf lo
```
Comment (dave-tucker) [inline src/start/dropping-packets.md:64]: The flag is `--iface`

File: src/start/dropping-packets.md [hunk 4]
```diff
@@ -41,7 +41,7 @@ We'll then call `block_ip` to determine the fate of the packet:
 
 In order to add addresses to block, we first need to get a reference to the `BLOCKLIST` map.
 Once we have it, it's simply a case of calling `blocklist.insert()`.
-We'll use the `IPv4Addr` type to represent our IP address as it's human-readable and can be easily converted to a `u32`. We'll block all traffic to `1.1.1.1` for this example.
+We'll use the `IPv4Addr` type to represent our IP address as it's human-readable and can be easily converted to a `u32`. We'll block all traffic **to** `192.168.0.10` sent over the `lo` interface, for this example.
```
Comment (dave-tucker) [inline src/start/dropping-packets.md:44]: I'm confused to the rationale of using `lo0`

The whole idea of providing the `--iface` flag was that the interface is configurable - e.g if you are using a VM and you don't have stable interface naming you could use `ens5` or whatever the OS decides you NIC is... or in my case, I used my wireless interface. `wlp2s0`. As you've somewhat eluded to here (by having to use `curl --interface lo`) there isn't much of interest happening on that interface anyway.

If it is not clear that the interface is configurable, I would suggest that we add a section at the beginning - perhaps in the environment setup or prerequisites.

something like;

```
In this example you will be attaching XDP programs to a network interface. The guide assumes you have an interface called `eth0`, but you may select any NIC on your machine.
```

This could be called out in the same block as the `XDP_FLAGS` note.

File: examples/myapp-03/myapp-ebpf/src/main.rs (hunk 1 above)
Comment (dave-tucker) [reply examples/myapp-03/myapp-ebpf/src/main.rs:71]: not just the code, but the prose and log output too.

File: src/start/dropping-packets.md [hunk 5]
```diff
@@ -61,18 +61,26 @@ We'll use the `IPv4Addr` type to represent our IP address as it's human-readable
...
+```
+
+### Testing the Allowed IP addresses
+
+```console
+curl --interface lo 192.168.0.22
+```
+
+```console
+LOG: SRC 192.168.0.22, ACTION 2
```
Comment (dave-tucker) [inline src/start/dropping-packets.md:74]: This is no longer `SRC` it's `DST` with your changes

**Review by dave-tucker — CHANGES_REQUESTED** [review]
Thanks for updating the template used to generate these examples @nitishm
I'd rather we didn't use the loopback interface as I'd like to remain as close to what you'd do in the "real world" as possible. Other than that, this LGTM.

**Comment by nitishm** [comment]
@dave-tucker PTAL - I updated the files based on the review

**Review by dave-tucker — APPROVED** [review]
LGTM. Thanks @nitishm

File: src/start/hello-xdp.md [hunk 6]
```diff
@@ -4,7 +4,11 @@
 
 While there are myriad trace points to attach to and program types to write we should start somewhere simple.
 
-XDP (eXpress Data Path) programs permit our eBPF program to make decisions about packets that have been received on the interface to which our program is attached. To keep things simple, we'll build a very simplistic firewall to permit or deny traffic.
+XDP (eXpress Data Path) programs permit our eBPF program to make decisions about packets that have been sent out on the interface to which our program is attached. To keep things simple, we'll build a very simplistic firewall to permit or deny traffic.
```
Comment (alessandrod) [inline src/start/hello-xdp.md:7]: XDP is inbound only, I think this needs to be phrased better

Comment (dave-tucker) [reply src/start/hello-xdp.md:7]: good catch! agreed.

Comment (nitishm) [reply src/start/hello-xdp.md:7]: Any suggestions on how I could rephrase this to indicate that we are dropping on destination address but on the egress path?

Comment (alessandrod) [reply src/start/hello-xdp.md:7]: I think the previous text was fine?

But I also think that more generally the example could be clearer. Right now it makes assumptions about the ip addresses assigned to eth0, and I had never seen curl's `--interface` option before, why is it needed in this case?. It would probably be good to explain a bit more how the network interface is configured, maybe showing the output of `ip addr show eth0`

File: src/start/dropping-packets.md [hunk 7]
```diff
@@ -69,7 +69,7 @@ sudo ./target/debug/myapp --path ./target/bpfel-unknown-none/debug/myapp --iface
 ### Testing the Allowed IP addresses
 
 ```console
-curl --interface eth0 192.168.0.22
+curl eth0 192.168.0.22
```
Comment (alessandrod) [inline src/start/dropping-packets.md:72]: This won't work, it's not a valid curl invocation. Why is passing the interface needed at all?

Comment (nitishm) [reply src/start/dropping-packets.md:72]: Ah shoot. My bad I meant to delete that part as well. Will amend and push

**Comment by dmitris** [comment]
@nitishm I also created https://github.com/aya-rs/book/issues/18 to sync the XDP example with the generated code, let me know if you want to incorporate it, or whether I should wait for your change to get merged to avoid a conflict.

**Comment by dave-tucker** [comment]
Closing since #18 merged to sync with latest template and the repo layout has changed significantly rendering this stale
